src/ManifoldDiffEq.jl

```diff
@@ -2,34 +2,697 @@ module ManifoldDiffEq
 
 using ManifoldsBase
 using Manifolds
+
+using LinearAlgebra
+
+import ConstructionBase: constructorof
+
+using Accessors: @set
+
+using SciMLBase: isinplace, promote_tspan, solve!
+
 using SciMLBase:
-    SciMLBase, AbstractODEProblem, AbstractODEFunction, NullParameters, promote_tspan
-import SciMLBase: build_solution
-using OrdinaryDiffEq
+    AbstractDiffEqOperator,
+    AbstractODEFunction,
+    AbstractODEProblem,
+    CallbackSet,
+    NullParameters,
+    ODEFunction,
+    ReturnCode,
+    SciMLBase
+
+import SciMLBase: alg_order, isadaptive, solution_new_retcode, solve
+
+using DiffEqBase:
+    DiffEqBase,
+    ODE_DEFAULT_NORM,
+    ODE_DEFAULT_ISOUTOFDOMAIN,
+    ODE_DEFAULT_UNSTABLE_CHECK,
+    ODE_DEFAULT_PROG_MESSAGE
+
+using OrdinaryDiffEqCore:
+    default_controller,
+    fsal_typeof,
+    gamma_default,
+    get_differential_vars,
+    handle_dt!,
+    initialize_callbacks!,
+    initialize_d_discontinuities,
+    initialize_saveat,
+    initialize_tstops,
+    isdtchangeable,
+    qmax_default,
+    qmin_default,
+    qsteady_max_default,
+    qsteady_min_default,
+    uses_uprev
+
+import OrdinaryDiffEqCore: perform_step!
 
-using OrdinaryDiffEq:
-    InterpolationData,
+using OrdinaryDiffEqCore:
+    DEOptions,
+    ODEIntegrator,
     OrdinaryDiffEqAlgorithm,
-    OrdinaryDiffEqAdaptiveAlgorithm,
-    OrdinaryDiffEqMutableCache,
     OrdinaryDiffEqConstantCache,
-    trivial_limiter!,
-    constvalue,
-    @muladd,
-    @unpack,
-    @cache,
-    @..
-
-import OrdinaryDiffEq: alg_cache, alg_order, initialize!, perform_step!
+    OrdinaryDiffEqCore,
+    OrdinaryDiffEqMutableCache
 
 using RecursiveArrayTools
 
+using SimpleUnPack
+
+"""
+    abstract type AbstractManifoldDiffEqAlgorithm end
+
+A subtype of `OrdinaryDiffEqAlgorithm` for manifold-aware algorithms.
+"""
+abstract type AbstractManifoldDiffEqAlgorithm <: OrdinaryDiffEqAlgorithm end
+
+isadaptive(::AbstractManifoldDiffEqAlgorithm) = false
+
+"""
+    abstract type AbstractManifoldDiffEqAdaptiveAlgorithm end
+
+Counterpart of `OrdinaryDiffEqAdaptiveAlgorithm`.
+"""
+abstract type AbstractManifoldDiffEqAdaptiveAlgorithm <: AbstractManifoldDiffEqAlgorithm end
+
+isadaptive(::AbstractManifoldDiffEqAdaptiveAlgorithm) = true
+
+"""
+    struct ManifoldODESolution end
+
+Counterpart of `SciMLBase.ODESolution`. It doesn't use the `N` parameter.
+"""
+mutable struct ManifoldODESolution{
+    T,
+    uType,
+    uType2,
+    DType,
+    tType,
+    rateType,
+    P,
+    A,
+    IType,
+    S,
+    AC<:Union{Nothing,Vector{Int}},
+    R,
+    O,
+}
+    u::uType
+    u_analytic::uType2
+    errors::DType
+    t::tType
+    k::rateType
+    prob::P
+    alg::A
+    interp::IType
+    dense::Bool
+    tslocation::Int
+    stats::S
+    alg_choice::AC
+    retcode::ReturnCode.T
+    resid::R
+    original::O
+end
+
+function ManifoldODESolution{T}(
+    u,
```

Comment:
```suggestion
    u::U,
```

with where U one can use that instead of the `typeof`. I find that a bit more readable. What do you think? (same for the others)

Reply:
With this many type parameters, `typeof` may actually be a bit more readable IMO.

Reply:
Ok.